src/main/java/com/example/auctionapp/Util/Search/QueryBuilder.java

```diff
@@ -0,0 +1,111 @@
+package com.example.auctionapp.Util.Search;
+
+import com.example.auctionapp.model.Product;
+
+import javax.persistence.EntityManager;
+import javax.persistence.TypedQuery;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Order;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryBuilder {
+
+    private final EntityManager entityManager;
+
+    private final CriteriaBuilder cb;
+
+    private final CriteriaQuery<Product> query;
+
+    private final Root<Product> resource;
+
+    private List<Predicate> predicates;
+
+    private Order order;
+
+    private Integer pageNumber;
+
+    private Integer pageSize;
+
+    public QueryBuilder(EntityManager entityManager) {
+        this.entityManager = entityManager;
+        this.cb = entityManager.getCriteriaBuilder();
+        this.query = this.cb.createQuery(Product.class);
+        this.resource = this.query.from(Product.class);
+        this.predicates = new ArrayList<>();
+    }
+
+    public void withName(String searchName) {
+        String pattern = "%" + searchName.toLowerCase() + "%";
+        Predicate name = cb.like(cb.lower(resource.get("name")), pattern);
+        predicates.add(name);
+    }
+
+    public void withSubcategory(Long subcategoryId) {
+        Predicate subcategory = cb.equal(resource.get("subcategory").get("id"), subcategoryId);
+        predicates.add(subcategory);
+    }
+
+    public void withColor(String color) {
+        Predicate colorPredicate = cb.equal(resource.get("color"), color);
```

Comment:
Use static string properties. In general, having string values in the actual method code is bad practice, they need to be extracted outside the usage.
This would lead to cleaner code.

```
private static String COLOR = "color"
Predicate colorPredicate = cb.equal(resource.get(COLOR), color);

```